cli/cmd/cluster/certificate-rotator.go

```diff
@@ -0,0 +1,227 @@
+package cluster
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/kinvolk/lokomotive/pkg/k8sutil"
+	"github.com/kinvolk/lokomotive/pkg/platform"
+)
+
+type certificateRotatorConfig struct {
+	clientSet            *kubernetes.Clientset
+	newCACert            string
+	logger               *log.Entry
+	daemonSetsToRestart  []platform.Workload
+	deploymentsToRestart []platform.Workload
+}
+
+type certificateRotator struct {
+	config certificateRotatorConfig
+}
+
+// CertificateRotateOptions contains the options for the RotateCertificates function.
+type CertificateRotateOptions struct {
+	Confirm    bool
+	Verbose    bool
+	ConfigPath string
+	ValuesPath string
+}
+
+// RotateCertificates replaces all certificates in a cluster.
+// due to the nature of it running as a lokoctl command it should be idempotency.
+func RotateCertificates(contextLogger *log.Entry, options CertificateRotateOptions) error {
+	cc := clusterConfig{
+		verbose:    options.Verbose,
+		configPath: options.ConfigPath,
+		valuesPath: options.ValuesPath,
+	}
+
+	c, err := cc.initialize(contextLogger)
+	if err != nil {
+		return fmt.Errorf("initializing: %w", err)
+	}
+
+	if err := canRotate(c); err != nil {
+		return fmt.Errorf("cannot rotate cluster: %w", err)
+	}
+
+	// Tainting certificates so they get rotated.
+	if err := c.taintCertificates(); err != nil {
+		rollbackErr := c.untaintCertificates()
```

Comment:
Hmm, if we run this here but not in the rest of the error paths, it will not work. Also, adding this `untaint` on the other error paths will probably not be enough, as terraform might have generated new certs and we need to have the old ones on disk for everything to go smooth as this didn't happen.

Am I missing something?

Sorry you added this, I was genuinely asking what happened in the error path, didn't want to imply to add it. I think just failing and document what happens on error seems simpler, but as you prefer.

If we are handling this case, though, we need to test the error path works